### DISCUSSION TITLE
What is the current state of remote pattern matching execution?

### DISCUSSION THREAD

**Discussion by Habush** [opencog/atomspace#2994] [Q&A] [answered]
Hi,

I want to replace the [atomspace-rpc](https://github.com/Habush/atomspace-rpc/) code I wrote a while ago where I used it for executing pattern matching queries on a remote atomspace using gRPC? I was wondering if something similar has been implemented in the atomspace and if it has `scheme` bindings (if not, I can code this part).

**Comment by linas** [comment, marked as answer]
Remote Atomspace pattern matching works, I think. I think there are unit tests, I think they pass. If there are bugs, I'll fix them. gRPC is not needed. There are even plug-ins for automatically distributing; but the plug-ins is more like proof-of-concept, they just show how, but are not all that powerful.

All this is formally documented on the wiki and in assorted READMEs. I will attempt to sketch this out here.

 * Wiki -- https://wiki.opencog.org/w/StorageNode  The relevant part is the `fetch-query` callback. In scheme, you'd say `(fetch-query (BindLink, etc.))` and it will run the query on the remote storage node (remote server) and return results to the local server. 
 * The example https://github.com/opencog/atomspace/blob/master/examples/atomspace/persist-query.scm shows how to use it. 

The next level of detail is this: suppose you have a local atomspace, and you are connected to a remote atomspace (using the StorageNode) and you now want to have that remote AtomSpace connected to N other servers, somewhere else. This can be done by opening N StorageNodes to each server. But how to distribute queries? Ah! Well, I do have an answer for that: there are several plug-ins, that distribute queries and replies. Next post will review this.

* FYI, CogServer now supports Websockets in addition to telnet interfaces, and you can get to the scheme, json and python shells with it (I plan to add a mini-datalog shell, and a quasi-subset-of-metta shell to it, as well.) Next post reviews the websockets.

**Reply by linas** [reply]
Websockets: Start a cogserver, then go to https://github.com/opencog/cogserver/tree/master/examples/websockets  and then click on demo.html and I think it will "just work".

**Reply by linas** [reply]
Proxies. -- these are StorageNodes that can do things like mirror/raid, buffer and cache, including going out over the net.  Look at the various `persist*scm` example demos for how these work. The [examples/atomspace/README](../blob/master/examples/atomspace/README.md) knows all.

Documentation here: https://wiki.opencog.org/w/ProxyNode

**Reply by Habush** [reply]
Okay. This looks great! I will get through the docs and try to move the current code (probably will do this by end of this week) to use the new api. I'll document any issues I face in this thread.
